apps/sync/js/crypto/fxsyncwebcrypto.js

```diff
@@ -0,0 +1,290 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+/* global
+  crypto,
+  KeyDerivation,
+  StringConversion
+*/
+
+/* exported
+  FxSyncWebCrypto
+*/
+
+// WebCrypto-based client for Firefox Sync.
+
+var FxSyncWebCrypto = (() => {
+
+  const HKDF_INFO_STR = 'identity.mozilla.com/picl/v1/oldsync';
+
+  var FxSyncWebCrypto = function() {
+    // Basic check for presence of WebCrypto.
+    if (!crypto || !crypto.subtle) {
+      throw new Error('This environment does not support WebCrypto');
+    }
+
+    this.mainSyncKey = null;
+    this.bulkKeyBundle = null;
+  };
+
+  var importKeyBundle = (aesKeyAB, hmacKeyAB) => {
+    var pAes = crypto.subtle.importKey('raw', aesKeyAB,
+                                       { name: 'AES-CBC', length: 256 },
+                                       true, [ 'encrypt', 'decrypt' ]);
+    var pHmac =  crypto.subtle.importKey('raw', hmacKeyAB,
+                                         { name: 'HMAC', hash: 'SHA-256' },
+                                         true, [ 'sign', 'verify' ]);
+    return Promise.all([pAes, pHmac]).then(results => {
+      return {
+        aes: results[0],
+        hmac: results[1]
+      };
+    }, () => {
+      // err might be a TypeError if aesKeyAB or hmacKeyAB was no ArrayBuffer.
+      throw new Error('Could not import key bundle.');
+    });
+  };
+
+  FxSyncWebCrypto.prototype._importKb = function(kBByteArray) {
```

Comment:
I'm not super familiar with Gaia conventions, but shouldn't this be all `Object.freeze`d to hell so I can't swizzle in my own snooping changes?